### PR TITLE
avoid reading address from size 0 span

### PR DESCRIPTION
**Pull request by vhvb1989**
fixes: https://github.com/Azure/azure-sdk-for-c/issues/1351

### CODE REVIEW

**Review by antkmsft — APPROVED** [review]

File: sdk/src/azure/core/az_http_response.c [hunk 1]
```diff
@@ -32,6 +32,11 @@ static AZ_NODISCARD bool _az_is_http_whitespace(uint8_t c)
 static AZ_NODISCARD az_result _az_get_digit(az_span* ref_span, uint8_t* save_here)
 {
 
+  if (az_span_size(*ref_span) == 0)
```
Comment (RickWinter) [inline sdk/src/azure/core/az_http_response.c:35]: Per the Embedded C model this should be a precondition check.  Is there a specific scenario which forces us to be more protective at runtime in this situation.

Comment (vhvb1989) [reply sdk/src/azure/core/az_http_response.c:35]: well, the value can change during runtime, that's why precondition is not enough for this cases.
We are getting an HTTP response from the server and reading/parsing it.  If the response, at some point becomes incomplete or full of digits only, we are making sure we don't read beyond span size.

File: cmake-modules/AddCmockaFunctions.cmake [hunk 2]
```diff
@@ -0,0 +1,98 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (ahsonkhan) [inline cmake-modules/AddCmockaFunctions.cmake:1]: Why was this cmake module file added/checked-in as part of the bug fix? Please make sure to keep your PRs specific to the issue being resolved.

File: sdk/src/azure/core/az_http_response.c [hunk 3]
```diff
@@ -158,6 +163,12 @@ AZ_NODISCARD az_result az_http_response_get_next_header(
     }
   }
 
+  if (az_span_size(ref_response->_internal.parser.remaining) == 0)
```
Comment (ahsonkhan) [inline sdk/src/azure/core/az_http_response.c:166]: A bug fix like this should be accompanied with unit tests. Please make sure to add the tests.

**Review by ahsonkhan — COMMENTED** [review]
Apologies for the late feedback, but it is blocking. Please avoid checking in untested code changes, or unrelated modifications.

File: sdk/src/azure/core/az_http_response.c [hunk 4]
```diff
@@ -158,6 +163,12 @@ AZ_NODISCARD az_result az_http_response_get_next_header(
...
+  {
+    // avoid reading address if span is size 0
+    return AZ_ERROR_UNEXPECTED_END;
```
Comment (ahsonkhan) [inline sdk/src/azure/core/az_http_response.c:169]: This need to be documented in the header file as a possible return value for `az_http_response_get_next_header`

File: cmake-modules/AddCmockaFunctions.cmake (hunk 2 above)
Comment (vhvb1989) [reply cmake-modules/AddCmockaFunctions.cmake:1]: shouldn't be part of this PR.  I will remove it

Comment (vhvb1989) [reply cmake-modules/AddCmockaFunctions.cmake:1]: https://github.com/Azure/azure-sdk-for-c/pull/1375

Comment (ahsonkhan) [reply cmake-modules/AddCmockaFunctions.cmake:1]: Reverted in https://github.com/Azure/azure-sdk-for-c/pull/1377

**Pull request merged**